plugin/pkg/auth/authenticator/token/keystone/keystone.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	"errors"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/auth/user"
+)
+
+// KeystoneAuthenticator contacts openstack keystone to validate user's credentials passed in the request.
+// The keystone endpoint is passed during apiserver startup
+type KeystoneAuthenticator struct {
+	configPath       string
+	apiCallValidator validator
+}
+
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+	//FIXME kfox1111: To enhance performance, cache admin token/client.
+	glog.V(3).Infof("Starting Keystone Token Auth.")
+	userid, projectid, roles, valid, err := keystoneAuthenticator.apiCallValidator.validate(token)
+	if err != nil {
+		return nil, false, err
+	}
+	extra := make(map[string][]string, 3)
+	extra["alpha.kubernetes.io/keystone_authn"] = []string{"true"}
+	extra["alpha.kubernetes.io/keystone_project_id"] = []string{projectid}
+	extra["alpha.kubernetes.io/keystone_roles"] = roles
```

Comment:
I understand the desire to avoid it.
I also don't think the existing subjects map well to the keystone primitives.

So I think the 4 options are:
 1.  Reimplement RBAC like functionality in a Keystone specific AuthZ module (seems bad to me)
 2. Add direct support to K8s RBAC to keystone primitives (kind of bad, but I don't think will be very much code)
 3. Come up with some other K8s abstractions that are similar to Keystone ones that are reasonable to map to, and then add support to RBAC for. Then its more loosely coupled. (seems good, but may take a while)
 4. Shoe horn Keystone things into K8s groups and try to use RBAC as is (Seems like some thing will work but may require some rbac changes to allow parsing out substrings out of group names to fully work. Seems bad)

So I'm mostly for 2 or 3. 3 is the cleanest. 2 I think is the most pragmatic.